File: apps/android/app/src/main/kotlin/app/trackevolution/ui/LayoutClass.kt
package app.trackevolution.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.trackevolution.ui.theme.LayoutTokens

/**
 * How much width the app has to lay out in (spec: NS-34).
 *
 * The unit of decision is the **window**, never the device or the form factor: a
 * tablet, a Fold opened flat, a phone in landscape and a Chromebook window all
 * land in the same three rules, and none of them is asked what kind of hardware
 * it is. The names and breakpoints are Material's window size classes — which is
 * also what the iOS port uses, in `App/DesignSystem/LayoutClass.swift`, so the
 * two clients cannot classify the same window differently.
 *
 * Deliberately **not** in `:core`. Every value here is a fact about a window
 * rather than about the domain, and `:core` may not depend on Android at all
 * (`checkNoAndroidDependency`); the two platforms pin the breakpoints with a
 * test each instead of sharing an implementation.
 *
 * **One divergence from iOS, deliberate and open.** iOS classifies an iPhone in
 * landscape as compact whatever its width, because UIKit hands it a `.compact`
 * horizontal size class — the OS saying "this is a one-column device". Android
 * has no equivalent signal, so a phone in landscape is classified by its width
 * and a 915dp-wide Pixel lands in [Expanded], the same class a small tablet
 * gets. That costs nothing while medium and expanded are the same layout, which
 * is the case for the whole of ticket 1. The two-pane shell (#216) is where it
 * becomes a decision: two panes on a 915 × 412dp window is a worse phone, so
 * that ticket has to either guard on height or read `smallestScreenWidthDp`,
 * which is the stable "is this a phone" signal on this platform.
 */
enum class LayoutClass {
    /**
     * Under 600dp. Today's phone layout, unchanged — this spec adds classes above
     * it and touches nothing in it.
     */
    Compact,

    /**
     * 600–839dp. The phone layout with its column capped and centred and its
     * grids filling by width. Half of a folded-open Fold lives here.
     */
    Medium,

    /** 840dp and up. Two-pane list-detail, and analysis beside the track map. */
    Expanded;

    /** Whether the phone layout applies as-is. */
    val isCompact: Boolean get() = this == Compact

    companion object {
        /**
         * The Material breakpoint between a phone layout and a capped one.
         *
         * Duplicated in `LayoutClass.MEDIUM_MIN_DP` on iOS, and pinned there by
         * `LayoutClassTests` as it is here by [app.trackevolution.ui.LayoutClassTest]
         * — a PR that changes one number and not the other fails the other
         * platform's test.
         */
        const val MEDIUM_MIN_DP: Int = 600

        /** The Material breakpoint at which a second pane earns its place. */
        const val EXPANDED_MIN_DP: Int = 840

        /** The class for a window of this width. The pure rule iOS mirrors. */
        fun ofWidth(widthDp: Int): LayoutClass = when {
            widthDp >= EXPANDED_MIN_DP -> Expanded
            widthDp >= MEDIUM_MIN_DP -> Medium
            else -> Compact
        }
    }
}

/**
 * The window's layout class and the width of the column content is being laid
 * out in.
 *
 * Two values rather than one because they answer different questions and change
 * at different times: the class is a fact about the *window* (and stays that, so
 * a narrow detail pane never redraws itself as a phone), while the content width
 * is a fact about the column the caller is inside — [PageColumn] narrows it when
 * it caps the page, and a pane will narrow it again.
 */
@Immutable
data class LayoutMetrics(
    val layoutClass: LayoutClass,
    /** The usable width of the current column, gutters already deducted. */
    val contentWidth: Dp,
) {
    /**
     * How many columns of at least [minimum] fit, with [gap] between them — the
     * native reading of the web's `repeat(auto-fill, minmax(<minimum>, 1fr))`.
     *
     * Never fewer than one: a column narrower than the minimum gets one item per
     * row and lets it be as narrow as it must be, exactly as CSS does.
     */
    fun columns(minimum: Dp, gap: Dp = 12.dp): Int {
        if (contentWidth <= 0.dp || minimum <= 0.dp) return 1
        return maxOf(1, ((contentWidth + gap) / (minimum + gap)).toInt())
    }

    /** The same metrics for a narrower column. */
    fun narrowedTo(width: Dp): LayoutMetrics = copy(contentWidth = minOf(contentWidth, width))

    /**
     * How wide a second column beside the page should be, or null for one column.
     *
     * The two pages that split — the event's analysis, the vehicle's selected
     * part — share this rule rather than each carrying its own pair of numbers,
     * and it is the same rule and the same numbers as iOS's
     * `LayoutMetrics.sideColumnWidth`.
     *
     * It reads [contentWidth], **never [layoutClass]**, and that is the point.
     * The class is a fact about the *window* and stays one, so a pane never
     * decides it is a phone — but "is there room here for two columns" is a
     * question about the container this page is in, and on a tablet in portrait
     * those two answers disagree: an expanded window, and a detail pane with
     * ~627dp of usable width. Splitting that gave the page column 300dp with a
     * chart column beside it, where iOS ran the same arithmetic and put its
     * column off the side of the screen.
     *
     * `null` below [LayoutClass.EXPANDED_MIN_DP] is the window's own threshold
     * applied one level down: a column too narrow to be a two-pane window is too
     * narrow to hold two columns of its own.
     */
    fun sideColumnWidth(fraction: Float, minimum: Dp, maximum: Dp): Dp? {
        if (contentWidth < LayoutClass.EXPANDED_MIN_DP.dp) return null
        return (contentWidth * fraction).coerceIn(minimum, maximum)
    }
}

/**
 * The window's layout metrics, published once by [ProvideLayoutMetrics].
 *
 * `static` because it changes rarely — a fold, a rotation, a resize — and every
 * screen reads it: an ordinary `compositionLocalOf` would track readers for a
 * value that is better off invalidating the whole subtree on the few occasions
 * it moves.
 */
val LocalLayoutMetrics = staticCompositionLocalOf {
    // A phone, until the window is measured. Nothing renders before it is, and
    // defaulting the other way would flash a two-pane layout.
    LayoutMetrics(LayoutClass.Compact, 0.dp)
}

/**
 * Measure the window and publish its layout class to everything below.
 *
 * The width comes from `Configuration.screenWidthDp`, which is a **deviation
 * from NS-34's letter** — the spec says `WindowSizeClass` from `androidx.window`'s
 * window metrics — taken for two reasons, and recorded here rather than in a
 * commit message the way NS-15 and NS-16 recorded theirs.
 *
 * The first is accuracy, and it runs the opposite way to the usual advice.
 * Window metrics report the window's whole bounds, decorations included; the
 * configuration reports it with the system decorations removed. Material
 * recommends the former because most apps let the system inset them. This app
 * does not: it draws edge to edge and insets its own content with
 * `systemBarsPadding()` in [app.trackevolution.SignedInScaffold], so the width a
 * layout actually receives is the decoration-free one — which is the number the
 * configuration already carries. Reading the raw bounds would classify a window
 * by space the content never gets.
 *
 * The second is that it is live, testable and free: `Configuration` recomposes on
 * every resize (a fold, Stage-Manager-style multi-window, rotation), it is what
 * Robolectric's `w840dp` qualifiers set, and it works in a `@Preview` — where an
 * `Activity`-based `WindowSizeClass` reads the *host* window and previews at
 * 400/700/1000dp would all report the same class.
 *
 * The breakpoints themselves are unchanged, so this classifies windows exactly
 * as `WindowSizeClass` would.
 */
@Composable
fun ProvideLayoutMetrics(content: @Composable () -> Unit) {
    val widthDp = LocalConfiguration.current.screenWidthDp
    val layoutClass = LayoutClass.ofWidth(widthDp)
    val metrics = LayoutMetrics(
        layoutClass = layoutClass,
        contentWidth = (widthDp.dp - pageGutter(layoutClass) * 2).coerceAtLeast(0.dp),
    )
    CompositionLocalProvider(LocalLayoutMetrics provides metrics) {
        content()
    }
}

/**
 * The page padding for a layout class.
 *
 * A phone keeps the 16dp the logbook screens have always used — 28dp of margin
 * on each side of a 412dp screen is most of a column. Once there is width to
 * spare the app uses the web's own `--page-gutter`, which is what makes a capped
 * page on a tablet sit the way the same page does in a browser beside it.
 *
 * The screens themselves still apply [PHONE_PAGE_GUTTER], as they have since
 * NS-26; [PageColumn] adds the difference above compact width, so no screen
 * needed editing and none of them can be left behind.
 */
fun pageGutter(layoutClass: LayoutClass): Dp =
    if (layoutClass.isCompact) PHONE_PAGE_GUTTER else LayoutTokens.PAGE_GUTTER

/** The gutter every logbook screen has used since NS-26. */
val PHONE_PAGE_GUTTER: Dp = 16.dp

/**
 * How narrow a card in an auto-filling grid may get — the web's
 * `.cards { grid-template-columns: repeat(auto-fill, minmax(280px, 1fr)) }`.
 *
 * A phone is narrower than this, so those grids are one column there and the
 * compact layout is unchanged by construction rather than by a branch.
 */
val CARD_GRID_MINIMUM: Dp = 280.dp

/**
 * Cards laid out across as many columns as the width allows — the web's
 * `repeat(auto-fill, minmax(280px, 1fr))` — inside a `LazyColumn`.
 *
 * The rows stay lazy, which the phone list already relied on: this chunks the
 * list into rows of [columns] and emits one item per row, rather than one per
 * card. At one column — which is where every phone lands, since a phone is
 * narrower than [CARD_GRID_MINIMUM] — a row holds exactly one full-width card
 * and the emitted list is item-for-item what it was.
 *
 * Rows are keyed by their first card, so at one column the keys are unchanged
 * too and a scroll position survives.
 */
fun <T> LazyListScope.cardGridItems(
    items: List<T>,
    columns: Int,
    key: (T) -> Any,
    gap: Dp = 12.dp,
    itemContent: @Composable (T) -> Unit,
) {
    val rows = items.chunked(maxOf(1, columns))
    items(rows, key = { row -> key(row.first()) }) { row ->
        Row(horizontalArrangement = Arrangement.spacedBy(gap)) {
            row.forEach { item ->
                Box(Modifier.weight(1f)) { itemContent(item) }
            }
            // Keeps a short last row's cards the width of the ones above rather
            // than stretching two cards across four columns.
            repeat(maxOf(1, columns) - row.size) { Spacer(Modifier.weight(1f)) }
        }
    }
}

/**
 * The logbook's content column: capped at `--page-max` and centred above phone
 * width, the way the web app's `.shell` is.
 *
 * Below the cap `widthIn(max = …)` is a no-op, so a phone renders exactly as it
 * did. The narrowed metrics are republished so a grid inside counts its columns
 * against the column it is actually in rather than against the window.
 */
@Composable
fun PageColumn(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    val metrics = LocalLayoutMetrics.current
    // The screens pad themselves by PHONE_PAGE_GUTTER, so only the difference is
    // added here. At compact width that difference is zero and this whole
    // wrapper is inert, which is what makes the phone layout unchanged by
    // construction rather than by a branch in ten screens.
    val extraGutter = pageGutter(metrics.layoutClass) - PHONE_PAGE_GUTTER
    Box(modifier.fillMaxSize(), contentAlignment = Alignment.TopCenter) {
        Box(
            Modifier
                .fillMaxSize()
                .widthIn(max = LayoutTokens.PAGE_MAX)
                .padding(horizontal = extraGutter),
        ) {
            CompositionLocalProvider(
                LocalLayoutMetrics provides metrics.narrowedTo(
                    LayoutTokens.PAGE_MAX - pageGutter(metrics.layoutClass) * 2,
                ),
            ) {
                content()
            }
        }
    }
}

File: apps/ios/App/DesignSystem/LayoutClass.swift
import SwiftUI

/// How much width the app has to lay out in (spec: NS-34).
///
/// The unit of decision is the **window**, never the device or the idiom: an
/// iPad in Split View, a phone in landscape and a Mac-sized Stage Manager window
/// all land in the same three rules, and none of them is asked what kind of
/// hardware it is. The names and the breakpoints are Material's window size
/// classes, which is what lets the Android port use the same two numbers —
/// `LayoutClass.kt` in `:app` — and lets a change to one fail the other's test.
///
/// Deliberately **not** in the Kit. Every value here is a fact about a window
/// rather than about the domain, and the Kit must stay free of UI frameworks;
/// the two platforms pin the breakpoints with a test each instead of sharing an
/// implementation.
enum LayoutClass: String, CaseIterable, Sendable {
    /// Under 600pt. Today's phone layout, unchanged — this spec adds classes
    /// above it and touches nothing in it.
    case compact
    /// 600–839pt. The phone layout with its column capped and centred, its grids
    /// filling by width, and its controls no longer stretching edge to edge.
    /// Half of a folded-open tablet and an iPad in portrait Split View live here.
    case medium
    /// 840pt and up. Two-pane list-detail, and analysis beside the track map.
    case expanded

    /// The Material breakpoint between a phone layout and a capped one.
    ///
    /// Named `_DP` on both platforms even though iOS counts in points: the value
    /// is the shared thing, and a Swift-flavoured name here would hide the fact
    /// that `MEDIUM_MIN_DP` in `ui/LayoutClass.kt` has to be the same number.
    static let MEDIUM_MIN_DP: CGFloat = 600
    /// The Material breakpoint at which a second pane earns its place.
    static let EXPANDED_MIN_DP: CGFloat = 840

    /// The class for a window of this width.
    ///
    /// The pure rule, and the one the Android port mirrors — the size class
    /// never reaches it.
    static func ofWidth(_ width: CGFloat) -> LayoutClass {
        if width >= EXPANDED_MIN_DP { return .expanded }
        if width >= MEDIUM_MIN_DP { return .medium }
        return .compact
    }

    /// The class for a window of this width in this horizontal size class.
    ///
    /// Both inputs are needed and neither is sufficient. The size class alone
    /// cannot tell a 13-inch iPad from a half-screen Split View pane on it —
    /// both are `.regular`. The width alone would call an iPhone 17 Pro Max in
    /// landscape (932pt) an expanded window and hand a phone in someone's hands
    /// a two-pane layout; UIKit already knows that device is a one-column
    /// proposition and says so with `.compact`, so a compact size class is the
    /// phone layout by definition, whatever the number.
    static func of(width: CGFloat, horizontalSizeClass: UserInterfaceSizeClass?) -> LayoutClass {
        if horizontalSizeClass == .compact { return .compact }
        return ofWidth(width)
    }

    /// Whether the phone layout applies as-is.
    var isCompact: Bool { self == .compact }
}

/// The window's layout class and the width of the column content is being laid
/// out in.
///
/// Two values rather than one because they answer different questions and change
/// at different times: the class is a fact about the *window* (and stays that,
/// so a narrow detail pane never redraws itself as a phone), while the content
/// width is a fact about the column the caller is inside — `TEPage` narrows it
/// when it caps the page, and a pane will narrow it again.
struct LayoutMetrics: Equatable, Sendable {
    var layoutClass: LayoutClass
    /// The usable width of the current column, gutters already deducted.
    var contentWidth: CGFloat

    /// How many columns of at least `minimum` points fit, with `gap` between
    /// them — the native reading of the web's
    /// `repeat(auto-fill, minmax(<minimum>, 1fr))`.
    ///
    /// Never fewer than one: a column narrower than the minimum gets one item
    /// per row and lets it be as narrow as it must be, exactly as CSS does.
    func columns(minimum: CGFloat, gap: CGFloat = TESpacing.gridGap) -> Int {
        guard contentWidth > 0, minimum > 0 else { return 1 }
        return max(1, Int((contentWidth + gap) / (minimum + gap)))
    }

    /// The same metrics for a narrower column.
    func narrowed(to width: CGFloat) -> LayoutMetrics {
        LayoutMetrics(layoutClass: layoutClass, contentWidth: min(contentWidth, width))
    }

    /// How wide a second column beside the page should be, or nil for one column.
    ///
    /// The three pages that split — the event's analysis, the track's lap
    /// compare, the vehicle's selected part — share this rule rather than each
    /// carrying its own pair of magic numbers, because the thing that goes wrong
    /// is the same on all three and went wrong on all three.
    ///
    /// It reads `contentWidth`, **never the layout class**, and that is the whole
    /// point. The class is a fact about the *window* and stays one, so a pane
    /// never decides it is a phone — but "is there room here for two columns" is
    /// a question about the container this page is actually in, and on an iPad in
    /// portrait with the sidebar showing those two answers disagree: an expanded
    /// window, and a detail pane with 627pt of usable width. Splitting that pane
    /// gave the side column 341pt of slot for the 445pt of content the window's
    /// width had asked for, and the difference ran off the side of the screen.
    ///
    /// `nil` below `EXPANDED_MIN_DP` is the same threshold the window uses to
    /// earn its second pane, applied one level down: a column too narrow to be a
    /// two-pane window is too narrow to hold two columns of its own.
    func sideColumnWidth(fraction: CGFloat, minimum: CGFloat, maximum: CGFloat) -> CGFloat? {
        guard contentWidth >= LayoutClass.EXPANDED_MIN_DP else { return nil }
        return min(max(contentWidth * fraction, minimum), maximum)
    }
}

private struct LayoutMetricsKey: EnvironmentKey {
    /// A phone, until `RootView` measures the window. Nothing renders before it
    /// does, and defaulting the other way would flash a two-pane layout.
    static let defaultValue = LayoutMetrics(layoutClass: .compact, contentWidth: 0)
}

extension EnvironmentValues {
    var layout: LayoutMetrics {
        get { self[LayoutMetricsKey.self] }
        set { self[LayoutMetricsKey.self] = newValue }
    }
}

extension View {
    /// Measure the window and publish its layout class to everything below.
    ///
    /// A `GeometryReader` rather than anything read once: Stage Manager, Split
    /// View and rotation all resize the window while the app is running, and a
    /// class captured at launch would be wrong for the rest of the session.
    /// (`onGeometryChange(for:)` would be tidier and is iOS 18; the deployment
    /// target is 17.)
    func measuringLayoutClass() -> some View {
        modifier(MeasureLayoutClass())
    }

    /// Republish the metrics for a container narrower than the window — a pane.
    ///
    /// The **class stays the window's**, which is the spec's rule and the right
    /// one: a pane must not decide it is a phone and start hiding things. What
    /// has to change is the *content width*, because everything that counts
    /// columns counts them against the column it is actually in. Without this a
    /// 360pt sidebar inherits the window's 1064pt content width and lays its
    /// track cards out three across, at 110pt each — which is how a list pane
    /// ends up far too narrow for what is in it.
    ///
    /// Measured rather than assumed: `NavigationSplitView` picks the sidebar's
    /// width itself, anywhere between the minimum and maximum it is offered, and
    /// the user can drag it.
    ///
    /// **It measures with a `GeometryReader`, so it is greedy**: it takes every
    /// point offered to it and reports that as its own size. Two consequences,
    /// both of which have bitten:
    ///
    /// - It goes **inside** a fixed frame, never outside one.
    ///   `column.frame(width: w).measuringPaneWidth()` puts the greedy reader
    ///   *around* the frame, so an `HStack` sees two fully flexible children and
    ///   splits the row in half — the fixed width stops deciding anything, and
    ///   whatever it was is either short of the half (a gap of dead background)
    ///   or over it (content off the side of the window). Write
    ///   `column.measuringPaneWidth().frame(width: w)`.
    /// - A pushed screen does not inherit it. Applied to a `NavigationStack`, it
    ///   publishes into the stack's own root and not into what
    ///   `navigationDestination` builds, so a detail screen went on reading the
    ///   *window's* width while being laid out in a 1025pt pane. `RootView`
    ///   therefore applies it to each destination, where the reader is inside the
    ///   container it is measuring and there is nothing left to inherit through.
    func measuringPaneWidth() -> some View {
        modifier(MeasurePaneWidth())
    }
}

private struct MeasureLayoutClass: ViewModifier {
    @Environment(\.horizontalSizeClass) private var horizontalSizeClass

    func body(content: Content) -> some View {
        GeometryReader { proxy in
            content.environment(\.layout, metrics(for: proxy.size.width))
        }
    }

    private func metrics(for width: CGFloat) -> LayoutMetrics {
        let layoutClass = LayoutClass.of(width: width, horizontalSizeClass: horizontalSizeClass)
        return LayoutMetrics(
            layoutClass: layoutClass,
            contentWidth: max(0, width - 2 * TESpacing.pageGutter(for: layoutClass))
        )
    }
}

private struct MeasurePaneWidth: ViewModifier {
    @Environment(\.layout) private var layout

    func body(content: Content) -> some View {
        GeometryReader { proxy in
            content.environment(
                \.layout,
                LayoutMetrics(
                    layoutClass: layout.layoutClass,
                    contentWidth: max(
                        0,
                        proxy.size.width - 2 * TESpacing.pageGutter(for: layout.layoutClass)
                    )
                )
            )
        }
    }
}
